### DISCUSSION TITLE
Enable Link-Time Optimization (LTO) and evaluate other optimizations

### DISCUSSION THREAD

**Discussion by zamazan4ik** [aws/aws-secretsmanager-agent#9] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) is not enabled. I suggest enabling it since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance.

I suggest enabling LTO only for the Release builds to not sacrifice the developers' experience during the working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `release-lto` profile where additionally to regular `release` optimizations LTO also will be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application.

After applying LTO I can suggest evaluating other optimization options like evaluating Profile-Guided Optimization (PGO) (more materials and benchmarks are available at my repo: https://github.com/zamazan4ik/awesome-pgo) and Post-Link Optimization (PLO) with tools like LLVM BOLT. However, it makes sense to evaluate them only if you are interested in achieving the "peak performance" for AWS Secrets Manager - not sure that it's a critical question right now.

Thank you.

**Comment by simonmarty** [comment]
Experimenting with this in #38

Stripping symbols and reducing codegen units reduces the binary size significantly. LTO actually appears to increase binary size.

We'll have to run load tests and see if this has a noticeable effect on performance.

**Reply by zamazan4ik** [reply]
I suggest you also add `lto = "fat"` to your experiments. Practical results from Fat LTO can differ from Thin LTO
